docs/spec/enums.rst

```diff
@@ -0,0 +1,342 @@
+behaviors::
+
+    class Color(Enum):
+        RED = 1
+        GREEN = 2
+        BLUE = 3
+
+    for color in Color:
+        reveal_type(color)  # Revealed type is 'Color'
+
+    reveal_type(Color["RED"])  # Revealed type is 'Literal[Color.RED]' (or 'Color')
+    reveal_type(Color(3))  # Revealed type is 'Literal[Color.BLUE]' (or 'Color')
+
+Calling an enum class does not invoke its constructor, as is normal
+in most Python classes. The defined ``__new__`` method of an enum class is
+replaced with a custom implementation that performs a value-based lookup of
+an enum member.
+
+Enum classes are immutable. Type checkers should generate an error if an enum
+is modified or deleted because these attempts will result in runtime
+errors::
+
+    Color.RED = 5  # Type checker error
+    del Color.RED  # Type checker error
+
+An Enum class with one or more defined members cannot be subclassed. They are
+implicitly "final". Type checkers should enforce this::
+
+    class EnumWithNoMembers(Enum):
+        pass
+
+    class Shape(EnumWithNoMembers):  # OK (because no members are defined)
+        SQUARE = 1
+        CIRCLE = 2
+
+    class ExtendedShape(Shape):  # Type checker error: Shape is implicitly final
+        TRIANGLE = 3
+
+
+Members and Non-Member Attributes
+---------------------------------
+
+When using the "class syntax", enum classes can define both members and
+non-member attributes. The ``EnumType`` metaclass applies a set of rules to
+distinguish between members and non-members. Type checkers should honor
+the most common of these rules. The lesser-used rules are optional. Some
+of these rules may be impossible to evaluate and enforce statically in cases 
+where dynamic values are used.
+
+* If an attribute is defined in the class body with a type annotation but
+  with no assigned value, a type checker should assume this is a non-member
+  attribute::
+
+    class Pet(Enum):
+        genus: str  # Non-member attribute
+        species: str  # Non-member attribute
+
+        CAT = 1  # Member attribute
+        DOG = 2  # Member attribute
+
+  Within a type stub, members can be defined using the actual runtime values,
+  or a placeholder of ``...`` can be used::
+
+    class Pet(Enum):
+        genus: str  # Non-member attribute
+        species: str  # Non-member attribute
+
+        CAT = ...  # Member attribute
+        DOG = ...  # Member attribute
+
+* Members defined within an enum class should not include explicit type
+  annotations. Type checkers should infer a literal type for all members.
+  A type checker should report an error if a type annotation is used
+  for an enum member because this type will be incorrect and misleading
+  to readers of the code::
+
+    class Pet(Enum):
+        CAT = 1  # OK
+        DOG: int = 2  # Type checker error
+
+    reveal_type(Pet.CAT)  # Revealed type is Literal[Pet.CAT]
+    reveal_type(Pet.DOG)  # Revealed type is Literal[Pet.DOG]
```

Comment:
Above, with `reveal_type(Color(3))` `Color` was allowed -- should `Pet` be allowed here?